plugins/source.go

```diff
@@ -155,39 +158,26 @@ func (p *SourcePlugin) Sync(ctx context.Context, spec specs.Source, res chan<- *
 		if table.Multiplex != nil {
 			clients = table.Multiplex(c)
 		}
-		// we call this here because we dont know when the following goroutine will be called and we do want an order
-		// of table by table
-		// totalClients := len(clients)
-		// newN, err := helpers.TryAcquireMax(ctx, goroutinesSem, int64(totalClients))
-		// if err != nil {
-		// 	p.logger.Error().Err(err).Msg("failed to TryAcquireMax semaphore. exiting")
-		// 	break
-		// }
-		// goroutinesSem.TryAcquire()
-		w.Add(1)
-		go func() {
-			defer w.Done()
-			wg := sync.WaitGroup{}
-			p.logger.Info().Str("table", table.Name).Msg("fetch start")
-			tableStartTime := time.Now()
-			totalTableResources := 0
-			for _, client := range clients {
-				client := client
-
-				// i := i
-				wg.Add(1)
-				go func() {
-					defer wg.Done()
-					// defer goroutinesSem.Release(1)
-					totalTableResources += table.Resolve(ctx, client, fetchTime, nil, res)
-				}()
+		for _, client := range clients {
+			client := client
+			wg.Add(1)
+			if err := goroutinesSem.Acquire(ctx, 1); err != nil {
+				// This mean context was cancelled
+				return err
 			}
-			wg.Wait()
-			totalResources += totalTableResources
-			p.logger.Info().Str("table", table.Name).Int("total_resources", totalTableResources).TimeDiff("duration", time.Now(), tableStartTime).Msg("fetch table finished")
-		}()
+			go func() {
+				defer wg.Done()
+				defer goroutinesSem.Release(1)
+				// TODO: prob introduce client.Identify() to be used in logs
+				tableStartTime := time.Now()
+				p.logger.Info().Str("table", table.Name).Msg("fetch start")
+				totalTableResources := table.Resolve(ctx, client, startTime, nil, res)
+				totalResources += totalTableResources
+				p.logger.Info().Str("table", table.Name).Int("total_resources", totalTableResources).TimeDiff("duration", time.Now(), tableStartTime).Msg("fetch table finished")
```

Comment:
`table` should be passed in to this function as a parameter, so we're not using the variable from the scope outside this goroutine

Reply:
I think this should be fine because I do `table := table` and `client := client` so this workarounds the shadow variable thing.

Reply:
Ah thanks, I missed those shadow declarations 👍